x-pack/plugins/security/server/routes/authentication.ts

```diff
@@ -0,0 +1,160 @@
+        `,
+          'text/html'
+        )
+      );
+    }
+  );
+
+  router.get(
+    {
+      path: '/api/security/authc/saml/capture-url-fragment.js',
+      validate: false,
+      options: { authRequired: false },
+    },
+    (context, request, response) => {
+      const currentBasePath = basePath.get(request);
+      return response.custom(
+        createCustomResourceResponse(
+          `
+          window.location.replace(
+            '${currentBasePath}/api/security/authc/saml/start?redirectURLFragment=' + encodeURIComponent(window.location.hash)
+          );
+        `,
+          'text/javascript'
+        )
+      );
+    }
+  );
+
+  router.get(
+    {
+      path: '/api/security/authc/saml/start',
+      validate: { query: schema.object({ redirectURLFragment: schema.string() }) },
+      options: { authRequired: false },
+    },
+    async (context, request, response) => {
+      try {
+        const authenticationResult = await authc.login(request, {
+          provider: 'saml',
+          value: {
+            step: SAMLLoginStep.RedirectURLFragmentCaptured,
+            redirectURLFragment: request.query.redirectURLFragment,
+          },
+        });
+
+        // When authenticating using SAML we _expect_ to redirect to the SAML Identity provider.
+        if (authenticationResult.redirected()) {
+          return response.redirected({ headers: { location: authenticationResult.redirectURL! } });
+        }
+
+        return response.unauthorized();
+      } catch (err) {
+        logger.error(err);
+        return response.internalError();
+      }
+    }
+  );
+
+  // Generate two identical routes with new and deprecated URL and issue a warning if route with
+  // deprecated URL is ever used.
+  for (const path of ['/api/security/authc/saml/callback', '/api/security/v1/saml']) {
+    router.post(
+      {
+        path,
+        validate: {
+          body: schema.object({
+            SAMLResponse: schema.string(),
+            RelayState: schema.maybe(schema.string()),
+          }),
+        },
+        options: { authRequired: false },
+      },
+      async (context, request, response) => {
+        try {
+          if (path === '/api/security/v1/saml') {
+            const currentBasePath = basePath.get(request);
+            logger.warn(
+              `The "${currentBasePath}${path}" URL is deprecated and will stop working in the next major version, please use "${currentBasePath}/api/security/authc/saml/callback" URL instead.`,
+              { tags: ['deprecation'] }
+            );
+          }
+
+          // When authenticating using SAML we _expect_ to redirect to the SAML Identity provider.
+          const authenticationResult = await authc.login(request, {
+            provider: 'saml',
+            value: {
+              step: SAMLLoginStep.SAMLResponseReceived,
+              samlResponse: request.body.SAMLResponse,
+              redirectURL: request.body.RelayState,
```

Comment:
__note:__ we may validate that URL is relative here assuming we keep transmitting it as it's